File: CS/DevExpressWebGardenReporting/ReportStorageWebExtension1.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;
using DevExpress.XtraReports.UI;

namespace DevExpressWebGardenReporting {
    public class ReportStorageWebExtension1 : DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension {
        public override bool IsValidUrl(string url) {
            return true;
        }

        public override string SetNewData(XtraReport report, string defaultUrl) {
            this.SetData(report, defaultUrl);
            return defaultUrl;
        }

        public override byte[] GetData(string url) {
            if(url == "Invoice")
                return SerializeReport(new InvoiceReport());

            url += url.EndsWith(".repx") ? "" : ".repx";
            using(var fileStream = File.OpenRead(HttpContext.Current.Server.MapPath("~\\Reports\\" + url))) {
                var report = XtraReport.FromStream(fileStream, true);
                return SerializeReport(report);
            }
        }

        public override Dictionary<string, string> GetUrls() {
            var dictionary = new Dictionary<string, string>();
            dictionary.Add("Invoice", "Invoice Report");
            var files = Directory.GetFiles(HttpContext.Current.Server.MapPath("~\\Reports\\"));
            foreach(var item in files.Where(x => x.Contains(".repx")).Select(x => x.Split('\\').Last())) {
                dictionary.Add(item, item);
            }
            return dictionary;
        }

        public override void SetData(DevExpress.XtraReports.UI.XtraReport report, string url) {
            url += url.EndsWith(".repx") ? "" : ".repx";
            using(var fileStream = File.OpenWrite(HttpContext.Current.Server.MapPath("~\\Reports\\" + url))) {
                report.SaveLayoutToXml(fileStream);
            }
        }

        public override bool CanSetData(string url) {
            return true;
        }

        byte[] SerializeReport(XtraReport report) {
            using(MemoryStream stream = new MemoryStream()) {
                report.SaveLayoutToXml(stream);
                stream.Position = 0;
                return stream.GetBuffer();
            }
        }
    }
}

File: VB/DevExpressWebGardenReporting/ReportStorageWebExtension1.vb
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Web
Imports DevExpress.XtraReports.UI

Namespace DevExpressWebGardenReporting

    Public Class ReportStorageWebExtension1
        Inherits DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension

        Public Overrides Function IsValidUrl(ByVal url As String) As Boolean
            Return True
        End Function

        Public Overrides Function SetNewData(ByVal report As XtraReport, ByVal defaultUrl As String) As String
            SetData(report, defaultUrl)
            Return defaultUrl
        End Function

        Public Overrides Function GetData(ByVal url As String) As Byte()
            If Equals(url, "Invoice") Then Return SerializeReport(New InvoiceReport())
            url += If(url.EndsWith(".repx"), "", ".repx")
            Using fileStream = File.OpenRead(HttpContext.Current.Server.MapPath("~\Reports\" & url))
                Dim report = XtraReport.FromStream(fileStream, True)
                Return SerializeReport(report)
            End Using
        End Function

        Public Overrides Function GetUrls() As Dictionary(Of String, String)
            Dim dictionary = New Dictionary(Of String, String)()
            dictionary.Add("Invoice", "Invoice Report")
            Dim files = Directory.GetFiles(HttpContext.Current.Server.MapPath("~\Reports\"))
            For Each item In files.Where(Function(x) x.Contains(".repx")).[Select](Function(x) x.Split("\"c).Last())
                dictionary.Add(item, item)
            Next

            Return dictionary
        End Function

        Public Overrides Sub SetData(ByVal report As XtraReport, ByVal url As String)
            url += If(url.EndsWith(".repx"), "", ".repx")
            Using fileStream = File.OpenWrite(HttpContext.Current.Server.MapPath("~\Reports\" & url))
                report.SaveLayoutToXml(fileStream)
            End Using
        End Sub

        Public Overrides Function CanSetData(ByVal url As String) As Boolean
            Return True
        End Function

        Private Function SerializeReport(ByVal report As XtraReport) As Byte()
            Using stream As MemoryStream = New MemoryStream()
                report.SaveLayoutToXml(stream)
                stream.Position = 0
                Return stream.GetBuffer()
            End Using
        End Function
    End Class
End Namespace
